.gitlab-ci.yml

```diff
@@ -321,16 +321,22 @@ integration:docker:
     - integration:builds
     - job: integration:deployment
       optional: true
-  image: docker:20.10.11
   services:
-    - docker:20.10.11-dind
+    - docker:20.10.16-dind
   variables:
     DOCKER_TLS_CERTDIR: "/certs"
-  before_script:
-    - docker info
+    FF_NETWORK_PER_BUILD: "true"
   script:
-    - image="$(docker load --input ./builds/*docker* | cut -d' ' -f3)"
-    - docker run "$image"
+    - docker info
+    - >
+      nix-shell --run $'
+        image="$(docker load --input ./builds/*docker* | cut -d\' \' -f3)"
+        PK_TEST_DOCKER_IMAGE=$image \
+        PK_TEST_COMMAND=docker \
+        PK_TEST_PLATFORM=docker \
+        PK_TEST_DATA_DIR=/builds/$CI_PROJECT_PATH/tmp \
```

Comment:
I don't think the additional env variable for `PK_TEST_DOCKER_IMAGE` as it adds more configuration.

I propose something like this:

```
PK_TEST_PLATFORM='docker'
PK_TEST_COMMAND="$(which docker) run \${options} someimage:sometag
```

The `someimage:sometag` is specified outside.

Now you need to use a simple utility function to "template" the options and args.

```
const template = (str, vars) => str.replace(
  /\${(.*?)}/g, 
  (_, p) => vars[p]
);

template(pkTestCommand, { options: '--x y z' })
```

Reply:
That way the `PK_TEST_COMMAND` remains an actual command, it's just that options can be provided in an interpolated fashion.

And this allows future applications of `PK_TEST_COMMAND` to the other platforms in case there are special options required too.

Now where should the options be "templated"? It should be done in your `tests/utils/exec.ts`, you can create a common function to do it. The `template` can be a useful utility function put into `tests/utils/utils.ts`. Might even be useful elsewhere too.

Reply:
You would need to make sure you're using `exec` and not `execFileSync`.

Reply:
I don't think that's going to work. when using `child_process.spawn` it takes (command, args, options). So the command will need to be `docker` on it's own, and the args need to be `['run', ...options, 'someimage:sometag', ...pkargs]`.

If we really need to get rid of the extra env variable, I can get the image id inside the test by execing `docker image ls -f "reference=polykey*" -q`

Reply:
Doesn't the spawn take an arbitrary string as the first parameter? You should be able to pass a full command into `spawn`. You just need to enable `shell: true`.

Reply:
![spawn](https://user-images.githubusercontent.com/640797/180921312-8ad0040b-4114-46e9-a9b2-9aa8dfec78fe.png)

Reply:
So it should work like `child_process.spawn('ls -l | cat', [], { shell: true });`

Reply:
BTW I just noticed `killSignal` is set to `SIGTERM`, that doesn't work on windows right? Seems strange that it would default to signal that has no effect on windows. Or maybe the docs is not up to date.

Reply:
Ok so some new ideas since we discussed:

1. Try using `execFile` and `spawn` with `shell: true` option, and see if it's possible to pass an arbitrary command as the first argument. And what the second argument of `args` is used for, possibly just concatenated arguments after the string?
2. This allows us to use `PK_TEST_COMMAND='ls -l | docker run'` or any arbitrary shell command string. Do note that on windows, this will be a `CMD` shell and not a bash/sh shell.
3. The `PK_TEST_COMMAND` originally was just a path to a file, now its a shell command. If it's a shell command, it should support the ability to interpolate a shell variable like: `PK_TEST_COMMAND='docker run $DOCKER_OPTIONS image:tag'`. Where you can then interpolate `$DOCKER_OPTIONS` during `execFile` or `spawn` by using the env option such as `{ env: { ...process.env, DOCKER_OPTIONS: '--a --b --c' } }`.
4. The `PK_TEST_PLATFORM` and `PK_TEST_COMMAND` should be coupled together as all or nothing. Meaning they can only either be both set, or both unset (`undefined`). This check should be done on the `jest.config.js`. This ensures that the user has to set both, the platform for how it is run, and the command for what is run. This means that `PK_TEST_PLATFORM` should not be defaulted to `os.platform()` instead it defaults to undefined.
5. The end result is that target based tests will end up with an intermediate shell program, where as regular testing without the 2 env variables will use the original way of executing PK. We have to be wary of any kill signals now, since for docker there could be 2 intermediate processes the shell and the docker's shell (if using a shell), but if doing a direct execution, then it should be just the shell itself.
6. As for the `PK_TEST_DATA_DIRECTORY`, that should be renamed to `PK_TEST_TMPDIR` so it's obvious that it's overwriting the `os.tmpdir()` being used bby the tests. But it does not overwrite `TMPDIR` normally, as we don't want to change the meaning of the `TMPDIR`.

Reply:
Doing some testing with `child_process.spawn`. Looks like 1, 2, 3 above work as expected and should be doable. I'll make the changes and do some more testing.

Reply:
Just ran all of the tests locally. It works with 1 minor difference breaking a single notification test. When running with `shell: false` a sing arg in the args array can contain spaces and be considered a single arg. In the problem test we have the following args;

```ts
        [
          'notifications',
          'send',
          nodesUtils.encodeNodeId(receiverId),
          'test message 3',
        ],
```

With `shell: false` the arg `test message 3` is considered a single arg. as if we passed it to the command like `"test message 3"`. However  with `shell: true` each word is treated as a seperate arg. so the command things the message is just `test`.

This can be fixed by setting the message to `'"test message 3"'` but this will break the test when it runs normally. Since it's a pretty niche problem affecting only 1 test I'll just replace the spaces in the message with `_`.

Just keep this difference in behaviour in mind for the future.

Reply:
tracking what's been done here,
- [X] 1.
- [X] 2.
- [X] 3.
- [ ] 4.
- [ ] 5.
- [X] 6.

Reply:
For 4, we have 1 case where we can't set both. That's for normal testing in the CI. This means we'd either have to;
1. allow PK_TEST_PLATFORM to be set to `linux` with  `PK_TEST_COMMAND` being undefined.
2. Have `runTestifPlatform` run by default if `PK_TEST_PLATFORM` is not set.

I'm leaning towards 2. here but I may need to remove `'linux` from every `runTestIfPlatform` usage.

Reply:
What case is that? I was thinking that `PK_TEST_PLATFORM` is only useful during targeted execution. Otherwise tests should work normally across all platforms.

Reply:
Resolving this. It has been addressed.